### PR TITLE
PNDA-1322: Test Master Dataset Health

### PR DESCRIPTION
**Pull request by Manasa-Sriram-ML**
# Problem Statement:
PNDA 1322: Test Master Dataset Health

# Analysis:
Test Gobblin health
Test Dataset creation
Test Quarantine data creation

# Approach:
Verify test topic data is available by checking Kafka health in console metrics
Verify Gobblin is running periodically as per the time interval set in CRON job
Verify that Kafka data pull job is successful using Gobblin logs
Verify Master and Quarantine dataset directories and files are created in HDFS using WebHDFS queries
Verify that master data and quarantine data, matches with, what is generated by Kafka test plugin

# Test details
UBUNTU-PICO-CDH
UBUNTU-STD-CDH
UBUNTU-PICO-HDP
UBUNTU-STD-HDP
RHEL-PICO-CDH
RHEL-STD-CDH
RHEL-PICO-HDP
RHEL-STD-HDP

### CODE REVIEW

**Comment by willis27** [comment]
Tested on both CDH and HDP. No metrics (gobblin.health and/or dataset.health) are reported to 'Metrics' view although I can see the dataset plugins running and data from 'avro.internal.testbot' are moved to both PNDA_Datasets and quarantine HDFS folders.
screen shot 2018-02-12 at 23 47 51
screen shot 2018-02-12 at 23 47 31

**Comment by Manasa-Sriram-ML** [comment]
Is the dataset plugin running?

**Comment by willis27** [comment]
I think so, as data have been moved around which is part of dataset plugin's logic?

**Comment by Manasa-Sriram-ML** [comment]
If Gobblin job is executed successfully then we will be seeing data moved around

**Comment by willis27** [comment]
Yes, I can confirm gobblin job running okay, and metrics are not visible from 'Metrics' View.

**Comment by Manasa-Sriram-ML** [comment]
Will check it by deploying a cluster in my setup and let you know

File: src/main/resources/plugins/dataset/TestbotPlugin.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+        self.gobblin_log_path = None
+        self.hdfs_url = None
+        self.elk_ip = None
+        self.metric_console = None
+        self.num_attempts = None
+        self.uri = None
+        self.query = None
+        self.master_dataset_dir = None
+        self.quarantine_dir = None
+	self.console_user = None
```
Comment (jeclarke) [inline src/main/resources/plugins/dataset/TestbotPlugin.py]: fix mixed tabs and spaces by removing tabs

File: src/main/resources/plugins/dataset/TestbotPlugin.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+        parser.add_argument('--data_service', default='http://10.0.1.20:7000', help='DATA_SERVICE IP ADDRESS')
+        parser.add_argument('--cluster_name', default='cluster_name', help="PNDA CLUSTER NAME")
+        parser.add_argument('--gobblin_log_path', default='/var/log/pnda/gobblin/gobblin-current.log',
+                            help="Gobblin Log path")
+        parser.add_argument('--elk_ip', default='localhost:9200', help="Elasticsearch IP address")
+        parser.add_argument('--metric_console', default='localhost:3123', help="Console metric IP address")
+        parser.add_argument('--num_attempts', default=10, help="retry count to check gobblin health")
+        parser.add_argument('--master_dataset_dir', default='/user/pnda/PNDA_datasets/datasets', help="dataset dir")
+        parser.add_argument('--quarantine_dir', default='/user/pnda/PNDA_datasets', help="quarantine dir")
+	parser.add_argument('--console_user', default='pnda', help="PNDA console username for PAM auth")
```
Comment (jeclarke) [inline src/main/resources/plugins/dataset/TestbotPlugin.py]: fix mixed tabs and spaces by removing tabs

File: src/main/resources/plugins/dataset/TestbotPlugin.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+            if r.status_code is not 200:
+                LOGGER.warning("Error while trying to access console by PAM auth")
+                return kafka_health
+            response = requests.get(url, cookies=r.cookies)
+            if response.status_code is 200:
+                kafka_health_dict = json.loads(response.text)
+                if kafka_health_dict['currentData']['value'] == 'OK':
+                    kafka_health = True
+        except Exception as e:
+            LOGGER.error("Exception while checking KAFKA health due to %s", e)
```
Comment (jeclarke) [inline src/main/resources/plugins/dataset/TestbotPlugin.py]: log exception stack traces with traceback

File: src/main/resources/plugins/dataset/TestbotPlugin.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                    # Calculating current_time, cron interval time in milliseconds
+                    current_milli_time = int(time.time() * 1000)
+                    ref_time = int(current_milli_time - (self.cron_interval * 60000))
+                    modified_milli_time = avro_file.get('modificationTime')
+
+                    # Checking whether the modified time lies in between before_time, now_time.
+                    if modified_milli_time in range(ref_time, current_milli_time):
+                        LOGGER.debug("File has been modified recently with in last cron interval time")
+                        avro_file_name = avro_file.get('pathSuffix')
+                        return avro_file_name
```
Comment (jeclarke) [inline src/main/resources/plugins/dataset/TestbotPlugin.py]: not keen on returns like this out of the middle of a function without a good reason to do so

**Review by jeclarke — CHANGES_REQUESTED** [review]
Some general points: exception printing, tabs and spaces and returning from in the middle of functions. Please have a look at improving the quality for these. Also please run pylint and fix up anything that it flags up. @willis27 is reviewing the functionality so no comment around that from me.

File: src/main/resources/plugins/dataset/TestbotPlugin.py (hunk 4 above)
Comment (Manasa-Sriram-ML) [reply src/main/resources/plugins/dataset/TestbotPlugin.py]: Sure. Will make required changes for all the comments provided.
Thank you!

**Comment by Manasa-Sriram-ML** [comment]
Requested changes have been made.
To handle the internal KAFKA topic data in HDFS, made required changes to add delete policy by age to the dataset which get's created.
Have ran pylint check and it has passed with a score of 9.9

**Comment by Manasa-Sriram-ML** [comment]
Since salt is already handling on setting a policy to testbot dataset, I have removed that functionality from the dataset plugin

**Comment by willis27** [comment]
Changed to PNDA-4100. Close this PR.

**Pull request closed without merging**